include/pybind11/stl.h

```diff
@@ -305,6 +314,58 @@ template<> struct type_caster<std::experimental::nullopt_t>
     : public void_caster<std::experimental::nullopt_t> {};
 #endif
 
+#if defined(PYBIND11_HAS_FILESYSTEM)
+template<> struct type_caster<std::filesystem::path> {
+
+    static handle cast(const std::filesystem::path& path, return_value_policy, handle) {
+        PyObject* py_str = [](const auto& w) {
+            using char_type = std::filesystem::path::value_type;
+            if constexpr (std::is_same_v<char_type, char>) {
+                return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
+            } else if constexpr (std::is_same_v<char_type, wchar_t>) {
+                return PyUnicode_FromWideChar(w.c_str(), ssize_t(w.size()));
+            }
+        } (path.native());
+        if (!py_str) {
+            return nullptr;
+        }
+        return module::import("pathlib").attr("Path")(reinterpret_steal<object>(py_str))
+               .release();
+    }
+
+    bool load(handle handle, bool) {
+        using char_type = std::filesystem::path::value_type;
+        if constexpr (std::is_same_v<char_type, char>) {
+            PyBytesObject* py_bytes = nullptr;
```

Comment:
You're supposed to use `PyObject*` even when you know you need a bytes object. It would help with the `reinterpret_cast` below.

Reply:
good point, this allows factoring out a bit more common code.